futures-util/src/io/copy_into.rs

```diff
@@ -3,75 +3,32 @@ use futures_core::task::{Context, Poll};
 use futures_io::{AsyncRead, AsyncWrite};
 use std::io;
 use std::pin::Pin;
+use super::{BufReader, CopyBufInto};
+use pin_utils::unsafe_pinned;
 
 /// Future for the [`copy_into`](super::AsyncReadExt::copy_into) method.
 #[derive(Debug)]
 #[must_use = "futures do nothing unless you `.await` or poll them"]
-pub struct CopyInto<'a, R: ?Sized + Unpin, W: ?Sized + Unpin> {
-    reader: &'a mut R,
-    read_done: bool,
-    writer: &'a mut W,
-    pos: usize,
-    cap: usize,
-    amt: u64,
-    buf: Box<[u8]>,
+pub struct CopyInto<R: AsyncRead, W> {
+    inner: CopyBufInto<BufReader<R>, W>,
 }
 
-impl<R: ?Sized + Unpin, W: ?Sized + Unpin> Unpin for CopyInto<'_, R, W> {}
+impl<R: AsyncRead, W> Unpin for CopyInto<R, W> where CopyBufInto<BufReader<R>, W>: Unpin {}
 
-impl<'a, R: ?Sized + Unpin, W: ?Sized + Unpin> CopyInto<'a, R, W> {
-    pub(super) fn new(reader: &'a mut R, writer: &'a mut W) -> Self {
+impl<R: AsyncRead, W> CopyInto<R, W> {
+    unsafe_pinned!(inner: CopyBufInto<BufReader<R>, W>);
+
+    pub(super) fn new(reader: R, writer: W) -> Self {
         CopyInto {
-            reader,
-            read_done: false,
-            writer,
-            amt: 0,
-            pos: 0,
-            cap: 0,
-            buf: Box::new([0; 2048]),
+            inner: CopyBufInto::new(BufReader::with_capacity(2048, reader), writer),
```

Comment:
I used the same buffer size here as what the old code used, I have no idea why this isn't just using `DEFAULT_BUF_SIZE` though...

Reply:
I don't know why it is used in old code, I think it should be preferable to use `DEFAULT_BUF_SIZE` [because `std::io::copy` uses `DEFAULT_BUF_SIZE`](https://github.com/rust-lang/rust/blob/00859e3e653973120006aaf3227823062dde1ba7/src/libstd/io/util.rs#L47) (I probably missed it [when defining `DEFAULT_BUF_SIZE`](https://github.com/rust-lang-nursery/futures-rs/pull/1573)).

Reply:
> I don't know why `2048` is used in old code

It may be related to `DEFAULT_BUF_SIZE` was once large: https://github.com/rust-lang/rust/pull/32695

Reply:
Changed it to just use the default.